packages/dds/tree/src/shared-tree/treeApi.ts

```diff
@@ -123,8 +171,51 @@ export const treeApi: TreeApi = {
 	},
 };
 
-function runTransaction(checkout: TreeCheckout, transaction: () => void | "rollback"): void {
+/**
+ * A requirement for a SharedTree transaction to succeed.
+ * @remarks Transaction constraints are useful for validating that the state of the tree meets some requirement when a transaction runs.
+ * In general, when running a transaction a client can validate their tree state in whatever way they wish and decide to either proceed with the transaction or not.
+ * However, they cannot know what the tree state will be when the transaction is _sequenced_.
+ * There may have been any number of edits from other clients that get sequenced before the transaction is eventually sequenced.
+ * Constraints provide a way to validate the tree state after the transaction has been sequenced and abort the transaction if the constraints are not met.
+ * All clients will validate the constraints of a transaction when it is sequenced, so all clients will agree on whether the transaction succeeds or not.
+ * @public
+ */
+export type Constraint = NodeExists;
+
+/**
+ * A transaction {@link Constraint | constraint} which requires that the given node exists in the tree.
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeExists {
```

Comment:
I wonder if we should call this `NodeInDocument` to avoid having this one-off definition of node "existence".

Reply:
I think that is a great idea, I'll do it.

Reply:
I renamed it to `NodeInDocumentConstraint`. I realized we probably want to put "constraint" at the end of the names for each of these so that we don't have to worry about colliding with some other random name in the future. I also tried doing this inside of a Constraint namespace but it messes up TSDoc and API extractor so I think this is the simplest. We expect people to make these objects in-line anyway (or with a helper function for more complicated ones) so they won't actually see the type name most of the time anyway.